File: agent-runtime/src/protocol.ts
import type { JsonObject, JsonValue } from "./types.js";

export const PROTOCOL_VERSION = 1 as const;

export const HOST_REQUEST_METHODS = [
  "runtime.initialize",
  "run.start",
  "run.steer",
  "run.cancel",
  "run.resume",
  "history.compact",
  "runtime.shutdown",
] as const;

export const RUNTIME_REQUEST_METHODS = ["tool.invoke", "host.log"] as const;

export const RUNTIME_EVENT_METHODS = [
  "run.started",
  "message.delta",
  "message.completed",
  "reasoning.delta",
  "steering.consumed",
  "tool.started",
  "tool.completed",
  "tool.failed",
  "interruption.requested",
  "usage.updated",
  "run.completed",
  "run.cancelled",
  "run.failed",
] as const;

export type HostRequestMethod = (typeof HOST_REQUEST_METHODS)[number];
export type RuntimeRequestMethod = (typeof RUNTIME_REQUEST_METHODS)[number];
export type RuntimeEventMethod = (typeof RUNTIME_EVENT_METHODS)[number];

type FrameBase = {
  jsonrpc: "2.0";
  protocolVersion: typeof PROTOCOL_VERSION;
  sessionId: string;
  runId: string;
  sequence: number;
};

export type RpcRequest = FrameBase & {
  id: string;
  method: HostRequestMethod | RuntimeRequestMethod;
  params: JsonObject;
};

export type RpcSuccess = FrameBase & {
  id: string;
  result: JsonValue;
};

export type RpcFailure = FrameBase & {
  id: string;
  error: { code: number; message: string; data?: JsonValue };
};

export type RpcResponse = RpcSuccess | RpcFailure;

export type RuntimeEvent = FrameBase & {
  eventId: string;
  method: RuntimeEventMethod;
  params: JsonObject;
};

export type RpcFrame = RpcRequest | RpcResponse | RuntimeEvent;

export type RuntimeFailureKind = "model_request" | "tool" | "runtime" | "unknown";

export type RuntimeFailureMetadata = {
  failureKind: RuntimeFailureKind;
  retryable: boolean;
  errorCode?: string;
};

export class ProtocolError extends Error {
  readonly code: number;
  readonly data?: JsonValue;

  constructor(code: number, message: string, data?: JsonValue) {
    super(message);
    this.name = "ProtocolError";
    this.code = code;
    if (data !== undefined) this.data = data;
  }
}

export function parseFrame(line: string): RpcFrame {
  let value: unknown;
  try {
    value = JSON.parse(line);
  } catch {
    throw new ProtocolError(-32700, "Invalid JSON");
  }
  if (!isRecord(value) || value.jsonrpc !== "2.0") {
    throw new ProtocolError(-32600, "Invalid JSON-RPC frame");
  }
  if (value.protocolVersion !== PROTOCOL_VERSION) {
    throw new ProtocolError(
      -32001,
      `Unsupported protocol version: ${String(value.protocolVersion)}`,
    );
  }
  requireString(value, "sessionId");
  requireString(value, "runId");
  if (!Number.isSafeInteger(value.sequence) || Number(value.sequence) < 0) {
    throw new ProtocolError(-32600, "Frame sequence must be a non-negative integer");
  }
  if ("method" in value) {
    requireString(value, "method");
    if ("eventId" in value) {
      requireString(value, "eventId");
    } else {
      requireString(value, "id");
    }
    if (!isRecord(value.params)) {
      throw new ProtocolError(-32602, "Frame params must be an object");
    }
  } else {
    requireString(value, "id");
    if (!("result" in value) && !("error" in value)) {
      throw new ProtocolError(-32600, "Response must contain result or error");
    }
  }
  return value as RpcFrame;
}

export function encodeFrame(frame: RpcFrame): string {
  return `${JSON.stringify(frame)}\n`;
}

export function isRequest(frame: RpcFrame): frame is RpcRequest {
  return "method" in frame && !("eventId" in frame);
}

export function isResponse(frame: RpcFrame): frame is RpcResponse {
  return !("method" in frame);
}

export function isEvent(frame: RpcFrame): frame is RuntimeEvent {
  return "method" in frame && "eventId" in frame;
}

export function runtimeFailureMetadata(data: JsonValue | undefined): RuntimeFailureMetadata {
  if (!isRecord(data)) return { failureKind: "unknown", retryable: false };
  const failureKind =
    data.failureKind === "model_request" ||
    data.failureKind === "tool" ||
    data.failureKind === "runtime" ||
    data.failureKind === "unknown"
      ? data.failureKind
      : "unknown";
  return {
    failureKind,
    retryable:
      failureKind !== "unknown" && typeof data.retryable === "boolean"
        ? data.retryable
        : false,
    ...(typeof data.errorCode === "string" ? { errorCode: data.errorCode } : {}),
  };
}

function requireString(value: Record<string, unknown>, key: string): void {
  if (typeof value[key] !== "string" || value[key] === "") {
    throw new ProtocolError(-32600, `Frame ${key} must be a non-empty string`);
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: src-tauri/src/agent_runtime/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct RpcFrame {
    pub jsonrpc: String,
    pub protocol_version: u32,
    pub session_id: String,
    pub run_id: String,
    pub sequence: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<RpcError>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RpcError {
    pub code: i64,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

impl RpcFrame {
    pub fn request(
        id: String,
        method: &str,
        session_id: &str,
        run_id: &str,
        sequence: i64,
        params: Value,
    ) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            protocol_version: PROTOCOL_VERSION,
            session_id: session_id.into(),
            run_id: run_id.into(),
            sequence,
            id: Some(id),
            event_id: None,
            method: Some(method.into()),
            params: Some(params),
            result: None,
            error: None,
        }
    }

    pub fn success(request: &Self, result: Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            protocol_version: PROTOCOL_VERSION,
            session_id: request.session_id.clone(),
            run_id: request.run_id.clone(),
            sequence: request.sequence,
            id: request.id.clone(),
            event_id: None,
            method: None,
            params: None,
            result: Some(result),
            error: None,
        }
    }

    pub fn failure(request: &Self, code: i64, message: impl Into<String>) -> Self {
        Self::failure_with_data(request, code, message, None)
    }

    pub fn failure_with_data(
        request: &Self,
        code: i64,
        message: impl Into<String>,
        data: Option<Value>,
    ) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            protocol_version: PROTOCOL_VERSION,
            session_id: request.session_id.clone(),
            run_id: request.run_id.clone(),
            sequence: request.sequence,
            id: request.id.clone(),
            event_id: None,
            method: None,
            params: None,
            result: None,
            error: Some(RpcError {
                code,
                message: message.into(),
                data,
            }),
        }
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.jsonrpc != "2.0" {
            return Err("unsupported JSON-RPC version".into());
        }
        if self.protocol_version != PROTOCOL_VERSION {
            return Err(format!(
                "unsupported agent protocol version {}",
                self.protocol_version
            ));
        }
        if self.sequence < 0 {
            return Err("sequence must be non-negative".into());
        }
        if self.session_id.is_empty() || self.run_id.is_empty() {
            return Err("sessionId and runId are required".into());
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unknown_protocol_versions() {
        let mut frame = RpcFrame::request(
            "1".into(),
            "run.start",
            "session",
            "run",
            0,
            serde_json::json!({}),
        );
        frame.protocol_version = 2;
        assert!(frame
            .validate()
            .unwrap_err()
            .contains("unsupported agent protocol"));
    }

    #[test]
    fn frame_uses_camel_case_wire_fields() {
        let value = serde_json::to_value(RpcFrame::request(
            "1".into(),
            "run.cancel",
            "s",
            "r",
            3,
            serde_json::json!({}),
        ))
        .unwrap();
        assert_eq!(value["protocolVersion"], 1);
        assert_eq!(value["sessionId"], "s");
    }

    #[test]
    fn failure_metadata_round_trips_in_error_data() {
        let request = RpcFrame::request(
            "1".into(),
            "tool.invoke",
            "s",
            "r",
            3,
            serde_json::json!({}),
        );
        let frame = RpcFrame::failure_with_data(
            &request,
            -32603,
            "Denied",
            Some(serde_json::json!({
                "failureKind": "tool",
                "retryable": false,
                "errorCode": "agent_path_denied"
            })),
        );
        let value = serde_json::to_value(frame).unwrap();

        assert_eq!(value["error"]["data"]["failureKind"], "tool");
        assert_eq!(value["error"]["data"]["retryable"], false);
        assert_eq!(value["error"]["data"]["errorCode"], "agent_path_denied");
    }
}
